File: epi_judge_java/epi/IsStringInMatrix.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;

public class IsStringInMatrix {
    private static int rows;
    private static int columns;
    private static int n;

    @EpiTest(testDataFile = "is_string_in_matrix.tsv")
    public static boolean isPatternContainedInGrid(List<List<Integer>> grid, List<Integer> pattern) {
        rows = grid.size();
        columns = grid.get(0).size();
        n = pattern.size();

        boolean[][][] deadEnds = new boolean[rows][columns][n];

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < columns; c++) {
                if (helperDfs(grid, r, c, pattern, 0, deadEnds)) {
                    return true;
                }
            }
        }

        return false;
    }

    private static boolean helperDfs(List<List<Integer>> grid, int r, int c, List<Integer> pattern, int index, boolean[][][] deadEnds) {
        if (pattern.size() == index) return true;

        if (r < 0 || r >= rows || c < 0 || c >= columns || deadEnds[r][c][index]) {
            return false;
        }

        if (grid.get(r).get(c).equals(pattern.get(index))
                && (helperDfs(grid, r - 1, c, pattern, index + 1, deadEnds)
                || helperDfs(grid, r + 1, c, pattern, index + 1, deadEnds)
                || helperDfs(grid, r, c - 1, pattern, index + 1, deadEnds)
                || helperDfs(grid, r, c + 1, pattern, index + 1, deadEnds))) {

            return true;
        }

        deadEnds[r][c][index] = true;
        return false;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IsStringInMatrix.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/is_string_in_matrix.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

const int DIRS[4][2] = {
    {-1, 0},
    {1, 0},
    {0, -1},
    {0, 1}
};

bool helper(const vector<vector<int> > &grid, int r, int c, const vector<int> &pattern, int idx) {
    if (grid[r][c] != pattern[idx]) return false;
    if (idx == pattern.size() - 1) return true;
    for (auto &[dr, dc]: DIRS) {
        int next_r = r + dr;
        int next_c = c + dc;

        if (next_r < 0 || next_r == grid.size() || next_c < 0 || next_c == grid[0].size()) continue;
        if (helper(grid, next_r, next_c, pattern, idx + 1)) {
            return true;
        }
    }
    return false;
}

bool IsPatternContainedInGrid(const vector<vector<int> > &grid, const vector<int> &pattern) {
    for (int i = 0; i < grid.size(); i++) {
        for (int j = 0; j < grid[0].size(); j++) {
            if (helper(grid, i, j, pattern, 0)) return true;
        }
    }
    return false;
}


int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"grid", "pattern"};
    return GenericTestMain(args, "is_string_in_matrix.cc",
                           "is_string_in_matrix.tsv", &IsPatternContainedInGrid,
                           DefaultComparator{}, param_names);
}
